base/linalg/generic.jl

```diff
@@ -103,21 +103,33 @@ function generic_vecnorm1(x)
     return convert(T, sum)
 end
 
+function generic_vecnorm2(n, scale, ssq, x)
+    @inbounds for i = 1:n
+        if x[i] != 0
+            absxi = norm(x[i])
+            if scale < absxi
+                ssq = 1 + ssq*(scale/absxi)^2
+                scale = absxi
+            else
+                ssq = ssq + (absxi/scale)^2
+            end
+        end
+    end
+    return scale * sqrt(ssq)
+end
+function generic_vecnorm2{T<:Number}(x::AbstractVector{T})
+    scale = zero(norm(zero(T)))
+    ssq = one(sqrt((abs2(norm(zero(T))) + abs2(norm(zero(T))))/2))
+    return generic_vecnorm2(length(x), scale, ssq, x)
+end
 function generic_vecnorm2(x)
-    maxabs = vecnormInf(x)
-    (maxabs == 0 || isinf(maxabs)) && return maxabs
-    s = start(x)
-    (v, s) = next(x, s)
-    T = typeof(maxabs)
-    scale::promote_type(Float64, T) = 1/maxabs
-    y = norm(v)*scale
-    sum::promote_type(Float64, T) = y*y
-    while !done(x, s)
-        (v, s) = next(x, s)
-        y = norm(v)*scale
-        sum += y*y
+    n = length(x)
+    if n == 0
+        throw(ArgumentError("input must either be an AbstractVector{T<:Number} or non-empty"))
     end
-    return convert(T, maxabs * sqrt(sum))
+    scale = zero(norm(x[1]))
```

Comment:
It's kind of annoying to compute `norm(x[1])` twice in this case.  Maybe add an `i0` argument to `generic_vecnorm2` and loop from `i = i0:n`.  That way you can pass `i0 == 2` to avoid recomputing the first element, which can be passed via `ssq`.